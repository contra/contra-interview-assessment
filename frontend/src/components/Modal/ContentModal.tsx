import React from 'react';
import styled from 'styled-components';
import Modal, { type ModalProps } from '.';

export type ContentModalProps = ModalProps & {
  text: string;
  title: string;
};

const ContentContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 12px;
`;

const Title = styled.h2`
  font-size: ${({ theme: { fontSize } }) => fontSize.larger};
  line-height: ${({ theme: { lineHeight } }) => lineHeight.large};
  color: ${({ theme: { colors } }) => colors.modalTitle};
`;

const Text = styled.p`
  font-size: ${({ theme: { fontSize } }) => fontSize.regular};
  line-height: ${({ theme: { lineHeight } }) => lineHeight.regular};
  color: ${({ theme: { colors } }) => colors.modalText};
`;

const ContentModal: React.FC<ContentModalProps> = ({
  text,
  title,
  children,
  ...baseModalProps
}) => {
  return (
    <Modal {...baseModalProps}>
      <ContentContainer>
        <Title>{title}</Title>
        <Text>{text}</Text>
      </ContentContainer>
      {children}
    </Modal>
  );
};

export default ContentModal;
