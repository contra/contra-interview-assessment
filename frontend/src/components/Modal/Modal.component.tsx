import { Fragment, MutableRefObject, useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import styles from './Modal.module.css';
import KeyboardManager from './keyboardManager';

export default function Modal({ isOpen, onClose = () => {}, children, focusRef }: { isOpen: boolean, onClose?: () => void, children?: JSX.Element, focusRef?: MutableRefObject<HTMLElement | null> }) {
  const modalRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const closeFunction = () => onClose();
    if (isOpen) {
      document.body.classList.add('stopScroll');
      KeyboardManager.register('Escape', closeFunction);
      focusRef?.current?.focus();
    }
    return () => {
      KeyboardManager.unregister('Escape', closeFunction)
      document.body.classList.remove('stopScroll');
    }
  }, [isOpen])

  return isOpen ? ReactDOM.createPortal(
    <Fragment>
      <div className={styles.overlay} onClick={onClose}>
      </div>
      <div className={styles.modalWrapper} ref={modalRef}>
        <div role="dialog" className={styles.modal}>
          {children}
        </div>
      </div>
    </Fragment>,
    document.body
  ) : null;
}
