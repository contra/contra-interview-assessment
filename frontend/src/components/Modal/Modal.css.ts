import { style, styleVariants } from '@vanilla-extract/css';
import { variables } from '@/theme.css';

const overlayStyle = style({
  alignItems: `center`,
  background: `rgba(0,0,0,0.1)`,
  display: `flex`,
  inset: 0,
  justifyContent: `center`,
  outline: `none`,
  overflow: `auto`,
  position: `fixed`,
  zIndex: 9_999,
});

export const overlayVariants = styleVariants({
  hidden: [overlayStyle, { display: `none` }],
  visible: [overlayStyle, { display: `flex` }],
});

export const modalWrapperStyle = style({
  '@media': {
    'screen and (min-width: 560px)': {
      padding: 0,
      width: `520px`,
    },
  },
  maxWidth: `calc(100vw - (${variables.spacing[3]} * 2))`,
  padding: variables.spacing[3],
  selectors: {
    [`${overlayVariants.visible}.active &`]: {
      transform: `scale(1)`,
    },
    [`${overlayVariants.visible}.inactive &`]: {
      transform: `scale(0.95)`,
    },
  },
  transform: `scale(1)`,
  transition: `transform 0.25s ease-in-out`,
  width: `100%`,
});

export const modalStyle = style({
  '@media': {
    'screen and (min-width: 560px)': {
      borderRadius: `2px`,
    },
  },
  background: variables.colors.white,
  backgroundClip: `padding-box`,
  boxShadow: `0 3px 6px -4px #0000001f, 0 6px 16px #00000014, 0 9px 28px 8px #0000000d`,
  display: `flex`,
  flexDirection: `column`,
  outline: `none`,
  width: `100%`,
});

export const headerStyle = style({
  background: variables.colors.white,
  borderBottom: `1px solid rgba(0,0,0,0.06)`,
  borderRadius: `2px 2px 0 0`,
  padding: `${variables.spacing[4]} 54px ${variables.spacing[4]} ${variables.spacing[6]}`,
  position: `relative`,
});

export const closeButtonStyle = style({
  ':focus': {
    color: variables.colors.slate[800],
  },
  ':hover': {
    color: variables.colors.slate[800],
    cursor: `pointer`,
  },
  alignItems: `center`,
  background: `none`,
  border: 0,
  bottom: 0,
  color: variables.colors.gray[300],
  display: `flex`,
  fontSize: variables.font.size['3xl'],
  justifyContent: `center`,
  padding: 0,
  position: `absolute`,
  right: 0,
  top: 0,
  transition: `color 0.3s ease-in-out`,
  width: `54px`,
});

export const bodyStyle = style({
  padding: variables.spacing[6],
});

export const footerStyle = style({
  alignItems: `center`,
  borderTop: `1px solid rgba(0,0,0,0.06)`,
  display: `flex`,
  gap: variables.spacing[2],
  justifyContent: `flex-end`,
  padding: `${variables.spacing['2.5']} ${variables.spacing[4]}`,
});
