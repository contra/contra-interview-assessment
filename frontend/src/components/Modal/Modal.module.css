.overlay, .modalWrapper {
  height: 100vh;
  width: 100vw;
  position: fixed;
  top: 0;
  left: 0;
}

.overlay {
  background-color: var(--overlay-bg-color);
  z-index: 1000;
}

/* 
  Seperating the modal from the overlay makes it easier to avoid mouse clicks
  potentially bubling to the overlay and triggering a close of the modal. 
*/
.modalWrapper {
  z-index: 1001;
  display: grid;
  align-items: center;
  justify-items: center;
  pointer-events:none;
}

.modal {
  background-color: var(--modal-bg-color);
  min-width: max(50vw, 500px);
  max-width: 90vw;
  max-height: 90vh;
  min-height: max(50vh, 200px);
  pointer-events: initial;
  overflow: auto;
}

/* 
  At a mobile size we will show the modal full screen
*/
@media (max-width: 500px) { 
  .modal {
    width: 100%;
    height: 100vh;
    min-width: initial;
    max-width: initial;
    min-height: initial;
    max-height: initial;
  }
}
