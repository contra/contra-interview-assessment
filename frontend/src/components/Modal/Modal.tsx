import { Icon } from '@iconify/react';
import React, {
  useEffect,
  useState,
  useMemo,
  type FC,
  useRef,
  useCallback,
} from 'react';
import { createPortal } from 'react-dom';
import { useSpring, animated } from 'react-spring';
import SimpleButton from '../SimpleButton';
import {
  bodyStyle,
  closeButtonStyle,
  headerStyle,
  footerStyle,
  modalWrapperStyle,
  modalStyle,
  overlayVariants,
} from './modal.css';

type PortalProps = {
  children?: React.ReactNode;
};

const Portal: FC<PortalProps> = ({ children }) => {
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
    return () => setMounted(false);
  }, []);

  return mounted ? createPortal(children, document.body) : null;
};

type ModalProps = {
  animate?: boolean;
  cancelButtonText?: string;
  children?: React.ReactNode;
  destroyOnClose?: boolean;
  escapable?: boolean;
  footer?: React.ReactNode;
  header?: React.ReactNode;
  lazy?: boolean;
  maskClosable?: boolean;
  okButtonText?: string;
  onCancel?: () => void;
  onOk?: () => void;
  onOpened?: () => void;
  title?: string;
  visible: boolean;
  width?: string;
};

const Modal: FC<ModalProps> = ({
  animate = true,
  title,
  visible,
  children,
  destroyOnClose = false,
  escapable = true,
  footer,
  header,
  lazy = false,
  maskClosable = true,
  okButtonText = `OK`,
  cancelButtonText = `Cancel`,
  onOk,
  onCancel,
  onOpened,
  width,
}: ModalProps) => {
  const id = `modal_${Date.now()}`;
  const [destroyed, setDestroyed] = useState(false);
  const [loaded, setLoaded] = useState(false);
  const [opened, setOpened] = useState(visible);
  const [maskVisible, setMaskVisible] = useState(visible);
  const [modalVisible, setModalVisible] = useState(visible);
  const modalElmRef = useRef<HTMLDivElement | null>(null);
  const maskElmRef = useRef<HTMLDivElement | null>(null);
  const label = useMemo(() => (title ? title : id), [id, title]);

  const overlayStyle = useSpring({
    config: {
      duration: animate ? 100 : 0,
    },
    onRest: () => {
      if (opened) {
        setModalVisible(true);
      } else {
        if (maskElmRef.current) {
          const level = Number.parseInt(
            maskElmRef.current.getAttribute(`aria-level`) ?? `0`,
            10
          );
          maskElmRef.current.style.zIndex = `9999`;
          maskElmRef.current.setAttribute(`aria-level`, `1`);
          if (level) {
            // get the previous modal and set it back to an active state
            const previousMask: HTMLElement | null = document.querySelector(
              `.${overlayVariants.visible.split(` `)[0]}.inactive[aria-level="${
                level - 1
              }"]`
            );
            if (previousMask) {
              previousMask.classList.remove(`inactive`);
              // reset focus to the currently active modal
              previousMask.focus();
            }
          }
        }

        setMaskVisible(false);

        if (destroyOnClose) {
          setDestroyed(true);
        }

        onCancel?.();
      }
    },
    onStart: () => {
      // set all other modals to an inactive state
      if (opened && maskElmRef.current) {
        Array.from(
          document.querySelectorAll(`.${overlayVariants.visible.split(` `)[0]}`)
        )
          .filter((elm) => elm !== maskElmRef.current)
          .forEach((elm) => elm.classList.add(`inactive`));
      }

      /* 
        make sure that no matter where the modal is placed in the dom that it shows up
        in the correct order in the stack when opened
      */
      const visibleMasks = Array.from(
        document.querySelectorAll(`.${overlayVariants.visible.split(` `)[0]}`)
      );
      if (visibleMasks.length > 1 && maskElmRef.current) {
        maskElmRef.current.setAttribute(`aria-level`, `${visibleMasks.length}`);
        const zIndexArray = visibleMasks.map((mask) => {
          const style = window.getComputedStyle(mask);
          return Number.parseInt(style.zIndex, 10);
        });
        zIndexArray.sort((a, b) => (a > b ? -1 : 1));
        maskElmRef.current.style.zIndex = zIndexArray[0]
          ? `${zIndexArray[0] + 1}`
          : `9999`;
      }
    },
    opacity: opened ? 1 : 0,
  });

  const modalAnimationStyle = useSpring({
    ...(modalVisible
      ? { opacity: 1, scale: 1, y: 0 }
      : { opacity: 0, scale: 0.9, y: 20 }),
    config: {
      duration: animate ? 200 : 0,
    },
    onRest: () => {
      if (modalElmRef.current) {
        modalElmRef.current.focus();
      }
    },
  });

  const closeModal = useCallback(() => {
    setModalVisible(false);
    setOpened(false);
  }, []);

  const onModalClick = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    event.stopPropagation();
  };

  useEffect(() => {
    setLoaded(true);
  }, []);

  useEffect(() => {
    if (loaded && onOpened && modalVisible) {
      setTimeout(onOpened, animate ? 250 : 0);
    }
  }, [onOpened, loaded, modalVisible, animate]);

  useEffect(() => {
    const maskElement = maskElmRef.current;
    const onEscape = (event: KeyboardEvent) => {
      event.stopPropagation();
      if (event.key === `Escape`) {
        closeModal();
      }
    };

    const onMaskClick = (event: MouseEvent) => {
      if (maskElement === event.target) {
        event.stopPropagation();
        closeModal();
      }
    };

    if (maskElement && escapable) {
      maskElement.addEventListener(`keydown`, onEscape);
    }

    if (maskElement && maskClosable) {
      maskElement.addEventListener(`click`, onMaskClick);
    }

    return () => {
      if (maskElement) {
        maskElement.removeEventListener(`keydown`, onEscape);
        maskElement.removeEventListener(`click`, onMaskClick);
      }
    };
  }, [maskElmRef, loaded, escapable, maskClosable, closeModal]);

  useEffect(() => {
    let firstElm: HTMLElement | null = null;
    let lastElm: HTMLElement | null = null;

    // trap the tab to the inside of the modal
    const onFirstElmKeyDown = (event: KeyboardEvent) => {
      if (event.key === `Tab` && event.shiftKey) {
        event.preventDefault();
        lastElm?.focus();
      }
    };

    const onLastElmKeyDown = (event: KeyboardEvent) => {
      if (event.key === `Tab` && !event.shiftKey) {
        event.preventDefault();
        firstElm?.focus();
      }
    };

    if (loaded && modalElmRef.current && modalVisible) {
      const elms = Array.from(
        modalElmRef.current.querySelectorAll(
          `select, input, textarea, button, a`
        )
      ).filter((elm) => {
        const style = window.getComputedStyle(elm);
        return (
          style.getPropertyValue(`display`) !== `none` &&
          style.getPropertyValue(`visibility`) !== `hidden`
        );
      });
      if (elms.length > 0) {
        firstElm = elms[0] as HTMLElement;
        lastElm = elms[elms.length - 1] as HTMLElement;

        firstElm.addEventListener(`keydown`, onFirstElmKeyDown);
        lastElm.addEventListener('keydown', onLastElmKeyDown);
      }
    }

    return () => {
      if (firstElm && lastElm) {
        firstElm.removeEventListener(`keydown`, onFirstElmKeyDown);
        lastElm.removeEventListener(`keydown`, onLastElmKeyDown);
      }
    };
  }, [modalElmRef, modalVisible, loaded]);

  useEffect(() => {
    if (visible) {
      setOpened(true);
      setDestroyed(false);
      setMaskVisible(true);
    } else {
      closeModal();
    }
  }, [visible, closeModal]);

  // lock body scrolling when modal is visible
  useEffect(() => {
    if (!loaded) return;
    document.body.style.overflowY = opened ? `hidden` : `auto`;
  }, [loaded, opened]);

  if (destroyed || (lazy && !visible)) return null;

  return (
    <Portal>
      <animated.div
        aria-level={1}
        className={`${
          overlayVariants[maskVisible ? 'visible' : 'hidden']
        } active`}
        data-testid="mask"
        ref={maskElmRef}
        style={overlayStyle}
        tabIndex={-1}
      >
        <div className={modalWrapperStyle} style={{ width }}>
          <animated.div
            aria-label={label}
            aria-modal
            className={modalStyle}
            id={id}
            onClick={onModalClick}
            ref={modalElmRef}
            role="dialog"
            style={modalAnimationStyle}
            tabIndex={0}
          >
            {header ?? (
              <div className={headerStyle}>
                {title}
                <button
                  aria-label="close"
                  className={closeButtonStyle}
                  onClick={closeModal}
                  type="button"
                >
                  <Icon icon="carbon:close" />
                </button>
              </div>
            )}
            <div className={bodyStyle}>{children}</div>
            {footer ?? (
              <div className={footerStyle}>
                <SimpleButton onClick={closeModal}>
                  {cancelButtonText}
                </SimpleButton>
                <SimpleButton onClick={onOk} type="primary">
                  {okButtonText}
                </SimpleButton>
              </div>
            )}
          </animated.div>
        </div>
      </animated.div>
    </Portal>
  );
};

export default Modal;
