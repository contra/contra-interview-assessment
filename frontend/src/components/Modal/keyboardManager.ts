
export class KeyboardManager {
  private subscribed = false;

  constructor() {
    this.subscribe();
  }

  private registeredKeys: { [key: string]:  ((e: KeyboardEvent) => void)[] } = {}

  private handleKey(event: KeyboardEvent) {
    let key = event.key;
    const keys = this.registeredKeys[key];

    if (keys && keys.length > 0) {
      const func = keys[keys.length - 1];
      func && func(event);
    }
  }

  subscribe() {
    if (typeof window !== 'undefined' && document && !this.subscribed) {
      this.subscribed = true;
      document.addEventListener('keyup', (e) => this.handleKey(e), false);
    }
  }

  unsubscribe() {
    document.removeEventListener('keyup', (e) => this.handleKey(e), false);
  }

  register(key: string, func: (e: KeyboardEvent) => void) {
    if (!this.subscribed) {
      this.subscribe();
    }
    if (!this.registeredKeys[key]) {
      this.registeredKeys[key] = [];
    }
    this.registeredKeys[key]?.push(func)
  }

  unregister(key: string, func: (e: KeyboardEvent) => void) {
    if (this.registeredKeys[key]) {
      this.registeredKeys[key] = this.registeredKeys[key]?.filter(f => f !== func) || [];
    }
  }
}

export default new KeyboardManager();