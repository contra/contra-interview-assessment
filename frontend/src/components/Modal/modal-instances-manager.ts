import { useEffect, useState } from 'react';

class ModalInstancesManager {
  // eslint-disable-next-line @typescript-eslint/prefer-readonly
  private modals: Map<string, number>;

  public constructor() {
    this.modals = new Map();
  }

  public add(modalId: string): number {
    this.modals.set(modalId, this.modals.size + 1);
    return this.modals.size;
  }

  public remove(modalId: string): void {
    this.modals.delete(modalId);
  }

  public isTopModal(modalId: string | null): boolean {
    if (!modalId) return false;
    return this.modals.get(modalId) === this.modals.size;
  }

  public hasActiveModals(): boolean {
    return Boolean(this.modals.size);
  }
}

export const manager = new ModalInstancesManager();

export const useModalManager = (modalId: string, isOpen?: boolean) => {
  const [index, setIndex] = useState(0);

  useEffect(() => {
    if (isOpen) {
      const modalIndex = manager.add(modalId);
      setIndex(modalIndex);
    }

    // eslint-disable-next-line consistent-return
    return () => {
      manager.remove(modalId);
      setIndex(0);
    };
  }, [isOpen, modalId]);

  return index;
};
