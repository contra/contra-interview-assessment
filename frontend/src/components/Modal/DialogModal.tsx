import React from 'react';
import styled from 'styled-components';
import Button from '../Button';
import ContentModal, { type ContentModalProps } from './ContentModal';

type Props = ContentModalProps & {
  cancelLabel: string;
  confirmLabel: string;
  onClickCancel: () => void;
  onClickConfirm: () => void;
  text: string;
  title: string;
};

const ButtonsContainer = styled.div`
  display: flex;
  gap: 16px;
  margin-top: 24px;
  justify-content: flex-end;
`;

const DialogModal: React.FC<Props> = ({
  cancelLabel,
  confirmLabel,
  onClickCancel,
  onClickConfirm,
  ...modalProps
}) => {
  return (
    <ContentModal {...modalProps}>
      <ButtonsContainer>
        <Button onClick={onClickCancel}>{cancelLabel}</Button>
        <Button onClick={onClickConfirm} variant="Secondary">
          {confirmLabel}
        </Button>
      </ButtonsContainer>
    </ContentModal>
  );
};

export default DialogModal;
