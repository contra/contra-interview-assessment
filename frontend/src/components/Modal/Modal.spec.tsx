import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Modal from './Modal.component';

describe('MyComponent', () => {
  beforeAll(() => {
    // @ts-ignore
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element;
    });
  });

  afterEach(() => {
    // @ts-ignore
    ReactDOM.createPortal.mockClear();
  });

  it('should render correctly', () => {
    const component = renderer.create(<Modal isOpen={true}><p>Hello World!</p></Modal>);

    expect(component.toJSON()).toMatchSnapshot();
  });
});