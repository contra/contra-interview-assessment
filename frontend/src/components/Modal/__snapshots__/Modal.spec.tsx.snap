// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent should render correctly 1`] = `
Array [
  <div
    className="overlay"
    onClick={[Function]}
  />,
  <div
    className="modalWrapper"
  >
    <div
      className="modal"
      role="dialog"
    >
      <p>
        Hello World!
      </p>
    </div>
  </div>,
]
`;
